lib/worker_manager.go

```diff
@@ -0,0 +1,295 @@
+	return h.id
+}
+
+func (h *tombstoneWorkerHandleImpl) IsTombStone() bool {
+	return true
+}
+
+func newWorkerManager(id MasterID, needWait bool, curEpoch Epoch) *workerManager {
+	return &workerManager{
+		needWaitForHeartBeat: needWait,
+		workerInfos:          make(map[WorkerID]*WorkerInfo),
+		masterEpoch:          curEpoch,
+		masterID:             id,
+	}
+}
+
+func (m *workerManager) Initialized(ctx context.Context) (bool, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !m.needWaitForHeartBeat {
+		return true, nil
+	}
+
+	if m.initStartTime.IsZero() {
+		m.initStartTime = time.Now()
+	}
+	if time.Since(m.initStartTime) > workerTimeoutDuration+workerTimeoutGracefulDuration {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (m *workerManager) Tick(ctx context.Context, router p2p.MessageRouter) (timedOutWorkers []*WorkerInfo) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// respond to worker heartbeats
+	for workerID, workerInfo := range m.workerInfos {
+		if !workerInfo.hasPendingHeartbeat {
+			if workerInfo.hasTimedOut() {
+				timedOutWorkers = append(timedOutWorkers, workerInfo)
+				delete(m.workerInfos, workerID)
+				m.tombstones[workerID] = &workerInfo.status
+			}
+			continue
+		}
+		reply := &HeartbeatPongMessage{
+			SendTime:  workerInfo.lastHeartBeatSendTime,
+			ReplyTime: time.Now(),
+			Epoch:     m.masterEpoch,
+		}
+		workerNodeID := workerInfo.NodeID
+		log.L().Debug("Sending heartbeat response to worker",
+			zap.String("worker-id", string(workerInfo.ID)),
+			zap.String("worker-node-id", workerNodeID),
+			zap.Any("message", reply))
+
+		msgClient := router.GetClient(workerNodeID)
+		if msgClient == nil {
+			// Retry on the next tick
+			continue
+		}
+		_, err := msgClient.TrySendMessage(ctx, HeartbeatPongTopic(m.masterID), reply)
+		if cerror.ErrPeerMessageSendTryAgain.Equal(err) {
+			log.L().Warn("Sending heartbeat response to worker failed: message client congested",
+				zap.String("worker-id", string(workerInfo.ID)),
+				zap.String("worker-node-id", workerNodeID),
+				zap.Any("message", reply))
+			// Retry on the next tick
+			continue
+		}
+		workerInfo.hasPendingHeartbeat = false
+	}
+	return nil
+}
+
+func (m *workerManager) HandleHeartBeat(msg *HeartbeatPingMessage) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	log.L().Debug("received heartbeat", zap.Any("msg", msg))
+	workerInfo, ok := m.workerInfos[msg.FromWorkerID]
+	if !ok {
+		log.L().Info("discarding heartbeat for non-existing worker",
+			zap.Any("msg", msg))
+		return
+	}
+	workerInfo.lastHeartBeatReceiveTime = time.Now()
+	workerInfo.lastHeartBeatSendTime = msg.SendTime
+	workerInfo.hasPendingHeartbeat = true
+}
+
+func (m *workerManager) UpdateWorkload(msg *WorkloadReportMessage) {
+	info, ok := m.getWorkerInfo(msg.WorkerID)
+	if !ok {
+		log.L().Info("received workload update for non-existing worker",
+			zap.String("master-id", string(m.masterID)),
+			zap.Any("msg", msg))
+		return
+	}
+	info.workload = msg.Workload
```

Comment:
The lock used in these two functions are not same, in `HandleHeartBeat`, `m.mu` protects worker info getter and worker info update. in `UpdateWorkload`, `m.mu` protects worker info getter only.